consensus_encoding/src/decode/decoders.rs

```diff
@@ -2,10 +2,135 @@
 
 //! Primitive decoders.
 
-use core::fmt;
+#[cfg(feature = "alloc")]
+use alloc::vec::Vec;
+#[cfg(feature = "alloc")]
+use core::convert::Infallible;
+use core::{fmt, mem};
+
+#[cfg(feature = "alloc")]
+use internals::write_err;
 
 use super::Decoder;
 
+/// Maximum size, in bytes, of a vector we are allowed to decode.
+#[cfg(feature = "alloc")]
+const MAX_VEC_SIZE: u64 = 4_000_000;
+
+/// A decoder that decodes a byte vector.
+///
+/// The encoding is expected to start with the number of encoded bytes (length prefix).
+#[cfg(feature = "alloc")]
+pub struct ByteVecDecoder {
+    prefix_decoder: Option<CompactSizeDecoder>,
```

Comment:
In 7e5ccd20d556e27ea1d5fbecdd41eaee88430acc is there a downside to eagerly initializing `prefix_decoder` in the `ByteVecDecoder::new` constructor and using it to represent the state of the `ByteVecDecoder`? If present, decoding the length prefix, else decoding the bytes. Then we could simplify the internal state and drop the `prefix_read` flag and perhaps simplify logic in the `min_bytes_needed` method.

Reply:
This may be possible, and other refactors. I'd like to get this merged and iterate on the internals (as well as adding more tests). I'm hoping to only fix grievous issues, it would be _really_ nice to get this in so we can rc.0 by Nashville.

Reply:
Ah, since I was doing the `PhantomData` I did this too. Thanks for pushing me.

Reply:
While doing so I forgot to re-set it after taking it out of the option and the tests did not fail (because single byte compact size). Goes with @apoelstra's comment below we need tests that use multi-byte compact size encodings.

Reply:
fwiw I still see the `prefix_decoder` in `ByteVecDecoder`, but looks like `VecDecoder<T: Decodable>` was updated to the new logic.

Reply:
Geez, hard to get good help huh. Thanks for being so patient with me.